File: pset6/sentimental-credit/credit.py
# TODO
import re


def main():

    try:
        card_num = input('Card number? ')
        check(card_num)
    except:
        print('INVALID')


def check(card_num):

    cardbrand = {  # Store credit card regular expressions inside a dictionary
        'VISA': '^[4]\d{12}$|^[4]\d{15}$',   # Visa regular expression
        'AMEX': '^34\d{13}$|^37\d{13}$',  # Amex regular expression
        'MASTERCARD': '^5[1-5]\d{14}$',  # Mastercard regular expression
    }

    card_num = re.sub("[ ]|[-]", '', card_num)  # Remove card number 'spaces' and '-'

    for key in cardbrand:
        compiler = re.compile(cardbrand[key])
        check_valid = re.match(compiler, card_num)  # Check for match according to cards regular expressions

        if check_valid != None:  # Matches
            checksum(card_num)  # Run Luhn’s Algorithm
            print(key)  # If valid checksum proceed to print the card brand
            return (0)
    else:
        exit(1)


def checksum(card_num):
    sum1 = 0
    sum2 = 0

    for x in range((len(card_num) - 2), -1, -2):    # Iterate through every other digit starting from second to last
        z = (int(card_num[x]) * 2)    # Multiply * 2 each digit
        sum1 += int(z % 10) + int(z / 10)   # Sum each digit from z

    for y in range((len(card_num) - 1), -1, -2):    # Iterate through every other digit starting from last
        sum2 += int(card_num[y])    # Sum the digits

    if ((sum1 + sum2) % 10) == 0:   # Check if sum1 + sum2 ends with 0
        return (0)
    else:
        exit(1)


if __name__ == '__main__':
    main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    int len, y, yy, h, k, kk, sum1 = 0, sum2 = 0, checksum = 0;
    char nstr[16];
    long x, xx, cnum;

    cnum = get_long("Credit card number: "); //get card number
    len = sprintf(nstr, "%ld", cnum); //number of chars
    for (y = 1; y < len; y += 2)
    {
        x = cnum / pow(10, y);   //Multiply every other digit by 2,
        k = (x % 10) * 2;   //starting with the number’s second-to-last digit
        h = k % 10 + (k / 10) % 10; //If k has 2 digits, sum them
        sum1 += h;
    }
    for (yy = 0; yy < len; yy += 2)
    {
        xx = cnum / pow(10, yy);   //sum of the digits that weren’t multiplied by 2
        kk = (xx % 10);
        sum2 += kk;
    }
    int start_cnum = cnum / pow(10, len - 2);   //card number first two digits
    if (((sum1 + sum2) % 10) == 0)
    {
        checksum = 1;
    }
    if (len == 16 && 50 < start_cnum && start_cnum < 56 && checksum == 1)
    {
        printf("MASTERCARD\n");
    }
    else if (len == 15 && checksum == 1 && (start_cnum == 34 || start_cnum == 37))
    {
        printf("AMEX\n");
    }
    else if (len < 17 && len > 12 && checksum == 1 && start_cnum >= 40 && start_cnum <= 49)
    {
        printf("VISA\n");
    }

    else
    {
        printf("INVALID\n");
    }
}